### ISSUE TITLE
Deploy workflow in Github Actions

### ISSUE THREAD

**Issue by oseasmoran73** [department-of-veterans-affairs/va.gov-team#29873]
## Details

It was noticed that if a newer commit finishes the pipeline before is predecessor and deploys, once its predecessor finishes it will deploy, thus not including the "latest" commit. This is due to race condition being present.

A few ideas on how to address this race condition could be:
- In the `deploy` step, we can check if the current `commit.sha` is the "latest" in master. If it's not, we can skip the deploy step
- We can remove `deploy` from `continuous-integration` workflow and make it it's own workflow. This could also potentially address #29578 depending on how it is laid out

## Acceptance Criteria

- [ ] Latest commit should always be deployed if CI succeeds in `master`
  - [ ] If new commit gets pushed while CI running in `master`, determine whether `deploy` step should execute
- [ ] Working as intended in vets-website
- [ ] Working as intended in content-build

**Comment by timwright12** [comment]
@ndsprinkle The current Jenkins implementation handles this by setting up a deployment queue. If there are any running master pipelines, the dev/staging deployment won't happen. 

Can we do the same thing here?

**Comment by ndsprinkle** [comment]
@timwright12 That's a good suggestion, I'll look into doing that. I'd essentially be checking to see if there are any running CI workflows?

**Comment by timwright12** [comment]
I'd have to think on it a bit more, but essentially when a CI completes on a commit, if there's a newer commit into master, don't deploy. if there are no newer commits, that means it's the latest and it can safely deploy.

Basically we don't want older commits deploying after newer ones because it would kill them off

**Comment by oseasmoran73** [comment]
@timwright12 I don't think that there is a way to have it in a "queue" system. The closest way (if possible) is for the deploy step to check if it is the latest one, if it is, safely deploy, otherwise, skip deploy step.

```
deploy:
  if: ${{github.sha == /ref/head/master/latest}}
```

something along the lines and it will then act like a "queue". This will ensure that no older commits deploy after newer ones. Thoughts?

**Comment by ndsprinkle** [comment]
Attempt at a summary of attempted methods so far:

`concurrency`:

Currently, we use `concurrency` in the continuous integration flow. It's purpose is to cancel in progress workflows when new commits are pushed to an already existing pull request. It is not used for any commits pushed/merged into master. 

`concurrency` not only is supported for an entire workflow, it is also supported for individual jobs. Our hope was to implement `concurrency` in the deploy job to cancel in progress deploys when a newer commit arrives at that point (or the opposite situation of cancelling an older commit workflow arriving at the deploy job **after** a newer one is already in progress). The issue with this approach is the same issue that spawned this ticket, which is that we have no way of identifying which workflow is newer or has newer commits once it has reached the deploy job. It would still be a race condition that would result in the faster workflow arriving at the deploy job first and the slower one cancelling the in progress deploy job or queuing behind it. 

`git rev-parse`:

One way we thought we could identify which workflow had newer commits was to use `git rev-parse`. We can use this command to output a SHA value that represents a commit. Different parameters can be passed in to `git rev-parse` to get different commit SHAs. 

In addition to this, Github actions has a built in `github` object that has a `sha` value and calling `github.sha` outputs the SHA of the current workflow. The theory was that we could compare this built in value to the SHA of origin/master `git rev-parse origin/master` to determine if the current workflow had the "newest" commits in origin/master. This comparison would happen at the deploy job using the built in `if` option that lets us use logical expressions to determine if the workflow should run.

While this sounded almost perfect in theory, the actual result was disappointing. This [pull request](https://github.com/department-of-veterans-affairs/vets-website/pull/18841) added SHA test outputs at various points throughout the continuous integration workflow. As seen in this [workflow](https://github.com/department-of-veterans-affairs/vets-website/runs/3704009311?check_suite_focus=true) under Archive (vagovdev)/Test Origin SHA with REVPARSE, using the command often results in errors thrown with git not even knowing what origin/master is. What's stranger still is the inconsistency, because in that same workflow under Build (vagovdev)/Test Origin SHA with REVPARSE, the output is successful and shows the proper SHA of origin/master.

More research will need to be done into this to figure out why it works sometimes in some jobs but not in others, but for now this is seeming like it won't work as a solution due to the inconsistencies with `git rev-parse` failing and with inconsistent SHA values that don't seem to line up with the actual commit SHAs.

`concurrency` (again):

To go full circle, if it is not possible to handle this at the deploy job, it should be possible to handle this race condition at the very top level of the entire workflow. As stated previously, the continuous integration workflow already uses `concurrency`, just not for master. One solution would be to enable this feature for master and have master specifically "queue" instead of "cancelling in progress". This would mean that "newer" continuous integration workflows would not start until the "older" commits have finished the entire workflow, which handles the race condition.

The main issue with this approach would be efficiency. If many commits are merged into master, things may start to back up since each workflow has to wait on the in progress workflow to finish, which is usually 30 minutes. We can get around this back up by not queuing at all and instead just cancelling the in progress workflow, but this may result in workflows never finishing/deploying because they are always being cancelled by newer commits. There also may be some tricky logic that makes sure `concurrency` works one way for master and another way for everything but master.

Even ignoring the potential side effects, it would still be ideal to handle this at the deploy step, so this solution should be seen as a somewhat last resort.

**Comment by timwright12** [comment]
Agreed that we should handle this in the deploy step or an added step right before the deploy (where ever we can get the commit hashes we need). I don't think we want to straight up cancel an in progress CI in master though. If there's a commit with a failing test we'll need to see where it fails in order to track it down properly. Cancelling works well in PRs because we just didn't want to clog up the system with running processes

**Comment by ndsprinkle** [comment]
@timwright12 After Oseas and I met yesterday, I think the plan is much closer to what you just described. There will be a new job right before the deploy job (and right after the archive job) that grabs the latest workflow run ID for CI. Then in the deploy job, we will compare that output to the built in `github` object that has the current run ID. If they are equal, then the deploy job will run. If not, it'll be skipped.

**Comment by ndsprinkle** [comment]
Probably a few more specific details to iron out and some potential messaging to add, but that's the general idea.

**Comment by ndsprinkle** [comment]
The PR created handles this situation in its most basic form. Developers merge PRs into master, that merge into master triggers the CI workflow, and the deploy job is conditionally run if the workflow is the latest CI workflow. To determine if the workflow is the "latest" workflow, a request is made to query for all the workflows that are the CI workflow in master and grabs the first object (newest based on ID) and uses the `run_number` property. If the `run_number` from the query is equal to the built in `github.run_number` of the running workflow, then the deploy job will run.

One unique situation that could have unexpected results is if a failed CI workflow is re-run. Workflows that are re-run do not have their ID or `run_number` updated. This means that a re-run workflow will never appear to be the "latest" workflow using the method in the PR. This may or may not matter depending on if we care about re-running failed CI workflows, because the next merged PR will be deployed over top of it anyway. If we do need to be able to re-run failed CI workflows, then we will need to add some logic that also checks `updated_at` timestamps or something along those lines to also determine if a workflow run is the latest.

The main other side effect I can see from this approach is that we may not be deploying as often as we'd like. CI workflows in master are taking ~30 minutes to complete, and if any other master CI workflow starts during that window, then the in progress job will not deploy. So if there are consistent merges into master throughout the day (spaced less than ~30 minutes apart), then nothing will deploy until the last merge of the day. I'd need to monitor this to see how frequently merges are made, but again I'm not sure how much this matters and if it will even be an issue.
